File: simple/matlab/randomwalkincome.m
function [Trans,xxx]=randomwalkincome(Z,sig)

%%%%%FUNCTION This function computes the value and the Markov matrix of a random income process 
%%%%% It takes as argument the possible number of values that the process
%%%%% can take and the standard deviation of the permanent income shock


%%%%%% vector
xxx=linspace(-3,3,Z);

pe=(xxx(end)-xxx(end-1))/2;

%%%%%INDIVIDUAL MATRIX
Trans=zeros(Z,Z);

for i=1:Z,
    for j=1:Z,
        diff=abs(xxx(i)-xxx(j));
        Trans(i,j)=double(normcdf(diff+pe,0,sig))-double(normcdf(diff-pe,0,sig));
    end;
end;

for i=1:Z,
    Trans(:,i)=Trans(:,i)./(sum(Trans(:,i)));
end;


end

File: simple/python/randomwalkincome.py
import numpy as np
from scipy.stats import norm 

def randomwalkincome(Z, sigma):
	'''value and matrix of Markov matrix

	Parameters
	----------
	Z : int
		number of possible values process can take
	sigma : float
		standard deviation of the permanent income shock

	Returns
	-------
	trans : array
		individual Markov matrix
	xxx : array
		vector
	'''

	xxx = np.linspace(-3, 3, Z)

	pe = (xxx[-1] - xxx[-2])/2

	trans = np.zeros((Z,Z))

	for i in xrange(Z):
		for j in xrange(Z):
			diff = abs(xxx[i] - xxx[j])
			trans[i,j] = norm.cdf(diff+pe, 0, sigma) - \
						 norm.cdf(diff-pe, 0, sigma)

	# trans /= trans.sum(axis=0)
	for i in xrange(Z):
		trans[:,i] /= trans[:,i].sum()

	return trans, xxx